apps/rush-lib/src/cli/utilities/pnpm/PnpmShrinkwrapFile.ts

```diff
@@ -0,0 +1,132 @@
+import * as os from 'os';
+
+import Utilities from '../../../utilities/Utilities';
+import { BaseShrinkwrapFile } from '../base/BaseShrinkwrapFile';
+
+interface IShrinkwrapDependencyJson {
+  /** Information about the resolved package */
+  resolution: {
+    /** The hash of the tarball, to ensure archive integrity */
+    integrity: string;
+    /** The name of the tarball, if this was from a TGX file */
+    tarball?: string;
+  };
+  /** The list of dependencies and the resolved version */
+  dependencies: { [dependency: string]: string };
+}
+
+/**
+ * This interface represents the raw shrinkwrap.YAML file
+ * Example:
+ *  {
+ *    "dependencies": {
+ *      "@rush-temp/project1": "file:./projects/project1.tgz"
+ *    },
+ *    "packages": {
+ *      "file:projects/library1.tgz": {
+ *        "dependencies: {
+ *          "markdown": "0.5.0"
+ *        },
+ *        "name": "@rush-temp/library1",
+ *        "resolution": {
+ *          "tarball": "file:projects/library1.tgz"
+ *        },
+ *        "version": "0.0.0"
+ *      },
+ *      "markdown/0.5.0": {
+ *        "resolution": {
+ *          "integrity": "sha1-KCBbVlqK51kt4gdGPWY33BgnIrI="
+ *        }
+ *      }
+ *    },
+ *    "registry": "http://localhost:4873/",
+ *    "shrinkwrapVersion": 3,
+ *    "specifiers": {
+ *      "@rush-temp/project1": "file:./projects/project1.tgz"
+ *    }
+ *  }
+ */
+interface IShrinkwrapYaml {
+  /** The list of resolved version numbers for direct dependencies */
+  dependencies: { [dependency: string]: string };
+  /** The description of the solved graph */
+  packages: { [dependencyVersion: string]: IShrinkwrapDependencyJson };
+  /** URL of the registry which was used */
+  registry: string;
+  /** The PNPM shrinkwrap internal versioning number */
+  shrinkwrapVersion: number | undefined;
+  /** The list of specifiers used to resolve direct dependency versions */
+  specifiers: { [dependency: string]: string };
+}
+
+export class PnpmShrinkwrapFile extends BaseShrinkwrapFile {
+  private _shrinkwrapJson: IShrinkwrapYaml;
+
+  public static loadFromFile(shrinkwrapYamlFilename: string): PnpmShrinkwrapFile | undefined {
+    try {
+      if (!fsx.existsSync(shrinkwrapYamlFilename)) {
+        return undefined; // file does not exist
+      }
+
+      // We don't use JsonFile/jju here because shrinkwrap.json is a special npm file format
+      // and typically very large, so we want to load it the same way that npm does.
+      const parsedData: IShrinkwrapYaml = yaml.safeLoad(fsx.readFileSync(shrinkwrapYamlFilename).toString());
+
+      return new PnpmShrinkwrapFile(parsedData);
+    } catch (error) {
+      throw new Error(`Error reading "${shrinkwrapYamlFilename}":` + os.EOL + `  ${error.message}`);
+    }
+  }
+
+  public getTempProjectNames(): ReadonlyArray<string> {
+    return this._getTempProjectNames(this._shrinkwrapJson.dependencies);
+  }
+
+  protected getTopLevelDependencyVersion(dependencyName: string): string | undefined {
```

Comment:
>protected [](start = 2, length = 9)

TypeScript lacks an "override" keyword.

When you override a method from the base class, I would add a comment such as "// override" or "// abstract" because otherwise it's super not obvious that this relationship exists.  We could also use the @microsoft/decorators for this purpose.